File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    int lowerBound(vector<vector<int>>& intervals, vector<int>& newInterval){
        int target=newInterval[0];
        int low=0;
        int high=intervals.size()-1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(intervals[mid][0]>=target){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return low;
    }
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        if(intervals.empty()){
            return {newInterval};
        }
        int lb=lowerBound(intervals,newInterval);
        intervals.insert(intervals.begin()+lb,newInterval);
        vector<vector<int>>ans;
        ans.push_back(intervals[0]);
        for(int i=1;i<intervals.size();i++){
            if(ans.back()[1]<intervals[i][0])ans.push_back(intervals[i]);
            else ans.back()[1]=max(ans.back()[1],intervals[i][1]);
        }
        return ans;
    }
};

File: 0057-insert-interval/0057-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]>list=new ArrayList<>();
        for(int[]interval:intervals){
            list.add(interval);
        }
        int low=0;
        int high=intervals.length-1;
        int target=newInterval[0];
        while(low<=high){
            int mid=low+(high-low)/2;
            if(intervals[mid][0]<=target){
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }
        list.add(low,newInterval);
        List<int[]>ans=new ArrayList<>();
        ans.add(list.get(0));
        int prev=list.get(0)[1];
        for(int i=1;i<list.size();i++){
            if(list.get(i)[0]<=prev){
                prev=Math.max(prev,list.get(i)[1]);
                ans.get(ans.size()-1)[1]=prev;
            }
            else{
                ans.add(list.get(i));
                prev=list.get(i)[1];
            }
        }
        return ans.toArray(new int[ans.size()][]);

    }
}